modules/apps/document-library/document-library-web/src/main/java/com/liferay/document/library/web/internal/servlet/CopyDLObjectServlet.java

```diff
@@ -0,0 +1,285 @@
+
+/**
+ * @author Jaime León
+ */
+@Component(
+	property = {
+		"osgi.http.whiteboard.servlet.name=com.liferay.document.library.web.internal.servlet.CopyDLObjectServlet",
+		"osgi.http.whiteboard.servlet.pattern=/copy_dl_object",
+		"servlet.init.httpMethods=POST"
+	},
+	service = Servlet.class
+)
+public class CopyDLObjectServlet extends HttpServlet {
+
+	@Override
+	protected void doPost(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws ServletException {
+
+		try {
+			String entryType = ParamUtil.getString(
+				httpServletRequest, "entryType");
+
+			if (entryType.equals(DLFolder.class.getSimpleName())) {
+				_copyFolder(httpServletRequest, httpServletResponse);
+			}
+			else if (entryType.equals(DLFileEntry.class.getSimpleName())) {
+				long fileShortcutId = ParamUtil.getLong(
+					httpServletRequest, "fileShortcutId");
+
+				if (fileShortcutId > 0) {
+					_copyFileShortcut(httpServletRequest, httpServletResponse);
+				}
+				else {
+					_copyFileEntry(httpServletRequest, httpServletResponse);
+				}
+			}
+		}
+		catch (IOException ioException) {
+			_log.error(ioException);
+
+			throw new ServletException(ioException);
+		}
+	}
+
+	@Override
+	protected void service(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException, ServletException {
+
+		try {
+			User user = _portal.getUser(httpServletRequest);
+
+			if ((user == null) || user.isGuestUser()) {
+				throw new PrincipalException.MustBeAuthenticated(
+					StringPool.BLANK);
+			}
+
+			PrincipalThreadLocal.setName(user.getUserId());
+
+			PermissionThreadLocal.setPermissionChecker(
+				_permissionCheckerFactory.create(user));
+
+			super.service(httpServletRequest, httpServletResponse);
+		}
+		catch (IOException | PortalException | ServletException exception) {
+			_log.error(exception);
+
+			_sendResponse(
+				httpServletResponse,
+				HttpServletResponse.SC_INTERNAL_SERVER_ERROR, null, null);
+		}
+	}
+
+	private void _checkDestinationRepository(long repositoryId)
+		throws PortalException {
+
+		Group group = _groupLocalService.fetchGroup(repositoryId);
+
+		if ((group != null) && group.isStaged() && !group.isStagingGroup()) {
+			throw new PortalException(
+				"cannot-copy-folders-to-the-live-version-of-a-group");
+		}
+	}
+
+	private void _copyFileEntry(
+			HttpServletRequest httpServletRequest,
+			HttpServletResponse httpServletResponse)
+		throws IOException {
+
+		long fileEntryId = ParamUtil.getLong(httpServletRequest, "fileEntryId");
+		long destinationFolderId = ParamUtil.getLong(
+			httpServletRequest, "destinationFolderId");
+		long destinationRepositoryId = ParamUtil.getLong(
+			httpServletRequest, "destinationRepositoryId");
+
+		try {
+			_checkDestinationRepository(destinationRepositoryId);
+
+			_dlAppService.copyFileEntry(
+				fileEntryId, destinationFolderId, destinationRepositoryId,
+				ServiceContextFactory.getInstance(
```

Comment:
This should be: 

ServiceContextFactory.getInstance(
	DLFileEntry.class.getName(), httpServletRequest));